### DISCUSSION TITLE
[Windows] WindowsInputPane memory leak

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#16900] [Q&A]
For each new window or not overlay popup a `WindowsInputPane` will be created but it never be collected form CG.

https://github.com/AvaloniaUI/Avalonia/blob/77fe0400fc57139c4a53fc0c7d148fca36c80cdd/src/Windows/Avalonia.Win32/WindowImpl.cs#L178

This is because the reference to `Windows Input Pane` is retauned by the following line:

https://github.com/AvaloniaUI/Avalonia/blob/77fe0400fc57139c4a53fc0c7d148fca36c80cdd/src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs#L87-L94

In order for the references present in `Handler` to be destroyed, it is necessary that there are no more native or managed references.

https://github.com/AvaloniaUI/Avalonia/blob/77fe0400fc57139c4a53fc0c7d148fca36c80cdd/src/Avalonia.MicroCom/CallbackBase.cs#L5-L56

The managed reference is destroyed immediately after creation.

https://github.com/AvaloniaUI/Avalonia/blob/77fe0400fc57139c4a53fc0c7d148fca36c80cdd/src/Windows/Avalonia.Win32/Input/WindowsInputPane.cs#L34-L39

While the native reference will never be removed because the `MicroCom` runtime invokes `OnUnreferencedFromNative` when the pointer's reference count reaches 0.

This does not happen because the `_cookie` pointer has the same value for all instances of `WindowsInputPane`

see following:

![Instances](https://github.com/user-attachments/assets/2b9e0202-1ed1-4354-9f23-f78cf3633f40)

![First](https://github.com/user-attachments/assets/6d5843f9-a80b-4871-abad-539c114d1b5a)

![Second](https://github.com/user-attachments/assets/dff8910d-713a-42f6-8de8-5f4662907f5c)

![thirth](https://github.com/user-attachments/assets/505a435c-e828-4baa-8f9b-128f91e886f4)

ecc..

How can we solve this problem? Make the `WindowsInputPane` instance creation Singleton?

>[!NOTE] 
> Before PR #16890, `WindowImpl` instances were also retained.

**Comment by workgroupengineering** [comment]
@kekekeks  Do you have any suggestions on how to fix this issue?

**Comment by maxkatz6** [comment]
InputPane itself should be disposed for any kind of toplevel from the WM_DESTROY https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs#L112

> This does not happen because the _cookie pointer has the same value for all instances of WindowsInputPane

That's inconvenient... This value should be unique identifier of registered handler.
We can, of course, use WeakReference inside of Handler object, but there must be another reason why this COM API returns the same registration identifier.

**Comment by MrJul** [comment]
The fix isn't needed anymore in Avalonia 12, please read https://github.com/AvaloniaUI/Avalonia/pull/16916#issuecomment-4741830457
